File: 100SameTree.cpp
/*
Given two binary trees, write a function to check if they are the same or not.

Two binary trees are considered the same if they are structurally identical and the nodes have the same value.

Example 1:

Input:     1         1
          / \       / \
         2   3     2   3

        [1,2,3],   [1,2,3]

Output: true
Example 2:

Input:     1         1
          /           \
         2             2

        [1,2],     [1,null,2]

Output: false
Example 3:

Input:     1         1
          / \       / \
         2   1     1   2

        [1,2,1],   [1,1,2]

Output: false
*/

class Solution{
public:
	bool isSameTree(TreeNode *p , TreeNode *q){
		//recursion
		if(!p && !q) return true;
		if(!p || !q) return false;
		return p->val == q->val && isSameTree(p->left, q->left) && isSameTree(p->right, q->right);

		//iteration
		/*
		stack<TreeNode*> s;
		s.push(p);
		s.push(q);

		while(!s.empty()){
			p = s.top() ; s.pop();
			q = s.top() ; s.pop();

			if(!p && !q) continue;
			if(!p || !q) return false;
			if(p->val != q->val) return false;

			s.push(p->left);
			s.push(q->left);

			s.push(p->right);
			s.push(q->right);
		}
		return ture;
		*/
	}
};

File: 100SameTree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isSameTree(self, p, q):
        """
        :type p: TreeNode
        :type q: TreeNode
        :rtype: bool
        """
        #recursion
        if p and q:
        	return p.val == q.val and self.isSameTree(p.left,q.left) and self.isSameTree(p.right,q.right)
        return p == q

        '''
		#iteration
        #use stack
		stack = [(p,q)]
		while stack:
			p,q = stack.pop()
			if p == None and q == None:
				continue
			if p == None or q == None:
				return False
			if p.val == q.val:
				stack.append((p.right, q.right))
				stack.append((p.left, q.left))
			else:
				return False
		return True
        '''

        '''
            #use queue
		queue = [(p,q)]
		while len(queue) != 0:
			p,q = queue.pop()
			if p == None and q == None:
				continue
			if p == None or q == None:
				return False
			if p.val == q.val:
				queue.insert(0, (p.left, q.left))
				queue.insert(0, (p.right, q.right))
			else:
				return Fasle
		return True
        '''
